### PR TITLE
doc: add notes that UDP receive functions must be called from same thread

### LINKED ISSUE #4993: doc: conn_udp: document that send and receives needs to be called from the same thread

### PR DESCRIPTION
**Pull request by OlegHahm**
fixes #4993.

### CODE REVIEW

File: sys/include/net/conn/udp.h [hunk 1]
```diff
@@ -51,6 +51,9 @@ typedef struct conn_udp conn_udp_t;
  * @param[in] family    The family of @p addr (see @ref net_af).
  * @param[in] port      The local UDP port for @p conn.
  *
+ * @note    @ref conn_udp_recvfrom needs to be called from the same thread as
+ *          for this function.
+ *
```
Comment (authmillenon) [inline sys/include/net/conn/udp.h]: Again: only true for GNRC. #3551 and #4713 don't have this problem.

File: sys/include/net/conn/udp.h [hunk 2]
```diff
@@ -93,6 +96,8 @@ int conn_udp_getlocaladdr(conn_udp_t *conn, void *addr, uint16_t *port);
  *
  * @note    Function may block.
  *
+ * @note    This function needs to be called from the same function as @ref conn_udp_create.
```
Comment (authmillenon) [inline sys/include/net/conn/udp.h]: Huh?!?! Why?

File: sys/posix/include/sys/socket.h [hunk 3]
```diff
@@ -398,6 +405,15 @@ ssize_t send(int socket, const void *buffer, size_t length, int flags);
...
  *                          to by the @p address argument.
  *
+ * @post    The socket will implicitely be bound to the unspecified address and
+ *          a random port, in case it is not already bound. Rationale: A client
+ *          should not require to explicitly call bind() to receive packets,
+ *          but is expected to receive replies sent to the ephemeral port that
+ *          was selected as a source port by the UDP implementation.
+ *
+ * @note    Any @ref recvfrom call that is called to receive an expected
+ *          response to this send command, must be called from the same thread.
```
Comment (authmillenon) [inline sys/posix/include/sys/socket.h]: See https://github.com/RIOT-OS/RIOT/pull/4995/files#r55391465

File: sys/include/net/conn/udp.h (hunk 1 above)
Comment (authmillenon) [inline sys/include/net/conn/udp.h]: I would rather add these in form of `@todo` notes since these are faulty behavior of `gnrc_conn_udp`/`gnrc_conn_ip`

Comment (OlegHahm) [inline sys/include/net/conn/udp.h]: This documentation describes the current state, #3551 and #4713 are not yet merged and so gnrc is the only available implementation of conn. Futhermore, I couldn't find documentation explicitly for gnrc_conn_udp.

File: sys/include/net/conn/udp.h (hunk 2 above)
Comment (OlegHahm) [inline sys/include/net/conn/udp.h]: Same as above. `conn_udp_create()` registers the thread for UDP.

File: sys/posix/include/sys/socket.h (hunk 3 above)
Comment (OlegHahm) [inline sys/posix/include/sys/socket.h]: See https://github.com/RIOT-OS/RIOT/pull/4995/files#r55393720

File: sys/include/net/conn/udp.h (hunk 2 above)
Comment (authmillenon) [inline sys/include/net/conn/udp.h]: Yeah, but the `conn` can be used from any other function in that thread (e.g. if `conn` is global or handed to other function via parameter).

File: sys/include/net/conn/udp.h (hunk 1 above)
Comment (authmillenon) [inline sys/include/net/conn/udp.h]: Just mention GNRC in the text. `conn` is conceptualized as *general* connection API. So regardless if other implementations exist or not: this is the current state in *GNRC*, and not an issue of this API.

Comment (authmillenon) [inline sys/include/net/conn/udp.h]: And again: I want to strongly advice to use `@todo` here.

Comment (authmillenon) [inline sys/include/net/conn/udp.h]: > Futhermore, I couldn't find documentation explicitly for gnrc_conn_udp.

there is none. It's just the implementation of this API for GNRC and it *should* not have different behavior than documented here ;-)

File: sys/include/net/conn/udp.h (hunk 2 above)
Comment (OlegHahm) [inline sys/include/net/conn/udp.h]: I don't know what you mean by "using conn", but if I call `conn_udp_recvfrom()` (which is basically a wrapper for `msg_receive()`) from a different thread than the one that got registered, I'm unlikely to receive anything, right?

Comment (authmillenon) [inline sys/include/net/conn/udp.h]: Yes... But you say "function" in the text above.

Comment (OlegHahm) [inline sys/include/net/conn/udp.h]: Yes, I don't understand the problem. I'm documenting `conn_udp_recvfrom()` which is a function.

Comment (OlegHahm) [inline sys/include/net/conn/udp.h]: Ah, now I saw the problem. You were to the second instance of the word `function`. That was simply a typo.

**Comment by OlegHahm** [comment]
addressed comments

**Comment by authmillenon** [comment]
Maybe mention that this is faulty behavior and needs to be fixed... Otherwise ACK, please squash.

**Comment by OlegHahm** [comment]
addressed comments and squashed

**Comment by immesys** [comment]
Oh! I did not know this. What happens if you use different threads? I'm pretty sure I have a different thread for sending and receiving. Maybe that explains some of my bugs.

**Comment by OlegHahm** [comment]
Well, the problem with gnrc_conn_udp() or the corresponding socket layer adaption is that the create function (conn_udp) respectively the sending function (POSIX) registers the thread at netreg. This means that only this thread will receive IPC messages, when a UDP datagram for the matching address and port (socket) is received. Otherwise these messages will get dropped.

**Pull request merged**